design-spec/priviledge-decentralized-swupdates.tex

```diff
@@ -177,32 +177,483 @@ \section{Design}
+    signal they readiness to run the new protocol. For instance, when a new
+    protocol requires the nodes to download and install new software, they must
+    have a way to determine when a sufficient majority of them have upgraded to
+    the new version.
+    % NOTE: might want to call these software and parameters protocol updates,
+    % however this seems to be too specific to the protocol. What happens if in
+    % the future we add the ability to upgrade the ledger rules dynamically?
+  \item[Non-signaled] An update is said to be \emph{non-signaled} if the nodes
+    can move to the new protocol version without needing to know if other nodes
+    are ready to move to said version as well. In this kind of updates we assume
+    the nodes are already running a version of the software that can switch to
+    the new protocol. An example of non-signaled protocol updates is
+    \emph{parameters-update}. In a parameters-update, the nodes have already the
+    logic for handling the parameters, and only the values of these parameters
+    change after an update is activated.
+  \end{description}
+\item[Non-protocol] These updates do not affect the consensus rules. The current
+  design does not restrict the nature of these updates. It is up to the ledger
+  layer to react to these updates accordingly. They might include:
+  \begin{itemize}
+  \item updates to the software that runs on the blockchain nodes, which does
+    not affect the protocol. Examples of such updates might include performance
+    or usability improvements.
+  \item transfer of funds. For instance an update proposal might involve funding
+    the development of a certain feature.
+  \item updates to components of the blockchain ecosystem.
+  \end{itemize}
+\end{description}
+
+Note that, according to the preceding definition, an update that requires the
+node to download and install new software to run a \emph{new protocol}, and that
+\emph{also} updates protocol parameters is considered a signaled protocol
+update.
+
+% TODO: mention somewhere that protocol updates must be activated one at a time,
+% whereas non-protocol updates can be approved in parallel.
 
 \subsection{The lifecycle of a decentralized update}
 \label{sec:phases-an-update}
 % FROM 3 The lifecycle of a decentralized software update
 
-% Why do we need to have a governance system that captures the whole life cycle
+The update protocol described in this document aims at capturing the whole life
+cycle of an update.
+% TODO: why do we need to have a governance system that captures the whole life cycle
 % of an update? So for instance, why not supporting update activations only?
-
-\subsection{Voting and delegation}
-\label{sec:voting-delegation}
-% FROM 4 Update governance
-
-\subsubsection{Delegation}
+%
+% Why distinguish phases at all? SIP and implementations can all be proposals.
+% They only differ in that some kinds of approved implementations can be
+% activated.
```

Comment:
Because the software update process is a logical sequence of decision steps. When you build software, you first have an idea, then you come up with a design, then move on to the implementation and finally you deploy and execute. This sequence of decisions is important to be identifiable in the stored events. Then only you can answer the WHEN, WHY and HOW the system has evolved the way it has. So this is the essential requirement. If you can achieve maintaining the logical sequence of events (i.e., decisions) with a single phase, then this is OK, I think.

Reply:
> Because the software update process is a logical sequence of decision steps. When you build software, you first have an idea, then you come up with a design, then move on to the implementation and finally you deploy and execute. This sequence of decisions is important to be identifiable in the stored events. 

So what do we need for this? I think simply:

- the notion of an update proposal
- a special class of "implementation proposals"
- a way to optionally link to other proposals.

This document would have been significantly shorter if we would not have had all the ideation/implemetation distinction. I think with the idea above we could have implemented these phases or any number of phases we wanted.

It is too late now, but ideally one should justify why each piece of complexity is added. If there's no good justification then this complexity is accidental (as opposed to essential).